lib/rubocop/cop/lint/out_of_range_ref_in_regexp.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # This cops looks for out of range referencing for Regexp, as while capturing groups out of
+      # out of range reference always returns nil.
+
+      # @example
+      #   /(foo)bar/ =~ 'foobar'\
+
+      #   # bad - always returns nil
+      #   puts $2 # => nil
+
+      #   # good
+      #   puts $1 # => foo
+      #
+      class OutOfRangeRefInRegexp < Cop
+        MSG = 'Do not use out of range reference for the Regexp.'
+
+        def investigate(processed_source)
+          ast = processed_source.ast
+          valid_ref = cop_config['Count']
+          ast.each_node do |node|
+            if node.regexp_type?
+              break if contain_non_literal?(node)
```

Comment:
I think you meant `next` instead of `break`. A test like the following would fail, right?
```
def foo
  /#{dynamic}/ =~ foo
end

def bar
  /(not)(dynamic)/ =~ bar
  $3 # << should raise error
end
```

Extracting this to a method might help make this more readable.